### ISSUE TITLE
NaN gradients in `RationalQuadraticSpline` for out-of-bounds inputs

### ISSUE THREAD

**Issue by cschoenle** [danielward27/flowjax#239]
> **Disclaimer**: I ran into this issue and then used an AI agent to analyze and debug it.

# Issue: Regression causing `NaN` gradients in `RationalQuadraticSpline` for out-of-bounds inputs

### Description
When using `RationalQuadraticSpline` (e.g. via coupling layers in normalizing flows), evaluating the flow on out-of-bounds inputs (like data from a newly discovered mode during iterative training) reliably produces `NaN` gradients. 

While the forward and inverse passes use `jnp.where(in_bounds, ...)` at the very end to correctly fallback to the identity mapping for out-of-bounds inputs, JAX still evaluates the mathematical operations on both branches. Because the inputs are no longer clipped or zeroed *before* the spline mathematics, operations evaluated outside the bounded `interval` can become numerically undefined.

For example, in `inverse_and_log_det`, an out-of-bounds `y` produces a negative value for the discriminant `b**2 - 4 * a * c`. The operation `jnp.sqrt` then evaluates to `NaN`. Similarly, in `transform_and_log_det`, the term `xi * (1 - xi)` becomes negative, which can cause the denominator `den` to equal zero, leading to division by zero.

These undefined operations produce `NaN` values that bypass the `jnp.where` condition during backpropagation (since the backward pass involves the VJP of `jnp.sqrt` or `/`, propagating `NaN` through the unselected branch), poisoning the gradients of the entire batch.

(see also the documentation of [jax.numpy.where](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.where.html))

### Historical Context (Regression in `v18.0.0`)
This issue was not present in version `17.2.0`. In older versions, there was explicit handling of out-of-bounds inputs *before* the math evaluation to explicitly prevent this JAX gradient propagation behavior. For example:
```python
in_bounds = jnp.logical_and(x >= self.interval[0], x <= self.interval[1])
x_robust = jnp.where(in_bounds, x, 0)  # To avoid nans
k = jnp.searchsorted(x_pos, x_robust) - 1
```

However, in commit `702d3e3f5` ("Use RealToIncreasingOnInterval and tidy up rqs"), the `x_robust` and `y_robust` variables (along with their `# To avoid nans` comments) were removed during refactoring. It appears it was assumed that the final `jnp.where` check at the end of the function would be sufficient, inadvertently exposing the flow to `NaN` gradients.

### Minimal Working Example (MWE)
```python
import jax
import jax.numpy as jnp
import equinox as eqx
from flowjax.bijections import RationalQuadraticSpline

# 1. Initialize RQS
rqs = RationalQuadraticSpline(knots=4, interval=1.0)

# 2. Add random noise to simulate a "trained" state where the
# parameters are perturbed from their exact uniform initialization.
params, static = eqx.partition(rqs, eqx.is_inexact_array)

key = jax.random.PRNGKey(42)
def add_noise(x):
    # nonlocal key  # Note: uncomment if running in a closure
    if x is not None:
        key, subkey = jax.random.split(key)
        return x + jax.random.normal(subkey, x.shape) * 1.5
    return x
    
params = jax.tree_util.tree_map(add_noise, params)
rqs_trained = eqx.combine(params, static)

# 3. Provide an out-of-bounds input for the inverse pass
y = jnp.array(50.0)

def loss(model, y_val):
    x, log_det = model.inverse_and_log_det(y_val)
    return x.sum()

# 4. Compute gradients
val, grad = eqx.filter_value_and_grad(loss)(rqs_trained, y)

# 5. Check for NaNs
import jax.tree_util as jtu
nan_found = any(
    jnp.any(jnp.isnan(leaf)) 
    for leaf in jtu.tree_leaves(grad) if eqx.is_array(leaf)
)
print(f"NaN in gradients: {nan_found}")  # Outputs True on >= v18.0.0
```

### Proposed Fix
The solution is to restore numerical safety by clipping `x` (and `y` for the inverse) to `self.interval` strictly *before* computing `k` and `xi`. The original unclipped variable should only be used in the final `jnp.where` fallback.

**In `transform_and_log_det`:**
```python
        in_bounds = jnp.logical_and(
            x >= jnp.array(self.interval[0]),
            x <= jnp.array(self.interval[1]),
        )
        x_clipped = jnp.clip(x, self.interval[0], self.interval[1])
        k = jnp.searchsorted(x_pos, x_clipped) - 1
        k = jnp.clip(k, min=0, max=len(x_pos) - 2)
        xi = (x_clipped - x_pos[k]) / (x_pos[k + 1] - x_pos[k])
        # ... rest of the math ...
        # (leave the final jnp.where(in_bounds, y, x) unchanged to use the unclipped x)
```

**In `inverse_and_log_det`:**
```python
        in_bounds = jnp.logical_and(
            y >= jnp.array(self.interval[0]),
            y <= jnp.array(self.interval[1]),
        )
        y_clipped = jnp.clip(y, self.interval[0], self.interval[1])
        k = jnp.searchsorted(y_pos, y_clipped) - 1
        k = jnp.clip(k, min=0, max=len(x_pos) - 2)
        # ... compute x ...
        # (leave the final jnp.where(in_bounds, x, y) unchanged to use the unclipped y)
```

**Comment by danielward27** [comment]
Oops! Yep, it's got the right idea I think! I think a nan in a where can result in the gradient even if it wouldn't be in the result. If you could do a pull request that would be great! Thanks for the report. If not, I'll probably get around to it in around a week or so, just away at the moment.

**Comment by danielward27** [comment]
Thanks for the report and fix @cschoenle!
